File: golang/binary_search.go
package main

type SearchResult struct {
	Index       int
	Comparisons int
}

type CompareFunc func(a, b int) int

func DefaultCompare(a, b int) int {
	return a - b
}

func BinarySearch(sortedArray []int, target int, compareFn CompareFunc) SearchResult {
	if compareFn == nil {
		compareFn = DefaultCompare
	}

	left := 0
	right := len(sortedArray) - 1
	comparisons := 0

	for left <= right {
		mid := left + (right-left)/2
		comparisons++
		comparison := compareFn(sortedArray[mid], target)

		if comparison == 0 {
			return SearchResult{Index: mid, Comparisons: comparisons}
		}

		if comparison < 0 {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return SearchResult{Index: -1, Comparisons: comparisons}
}

File: c/binary_search.c
#include "binary_search.h"

int default_compare(int a, int b) {
    return a - b;
}

SearchResult binary_search(
    const int *sorted_array,
    size_t length,
    int target,
    CompareFunc compare_fn
) {
    if (compare_fn == NULL) {
        compare_fn = default_compare;
    }

    int left = 0;
    int right = (int)length - 1;
    int comparisons = 0;

    while (left <= right) {
        int mid = left + (right - left) / 2;
        comparisons++;
        int comparison = compare_fn(sorted_array[mid], target);

        if (comparison == 0) {
            return (SearchResult){ .index = mid, .comparisons = comparisons };
        }

        if (comparison < 0) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return (SearchResult){ .index = -1, .comparisons = comparisons };
}
